File: src/server/api/routes/boards.ts
import { Hono, type Context } from "hono";
import { z } from "zod";
import { zValidator } from "@hono/zod-validator";

import { boardRoles } from "~/server/db/schema";
import {
  addBoardMember,
  createBoard,
  deleteBoard,
  duplicateBoard,
  getBoardDetail,
  listBoardMembers,
  listBoardsForUser,
  removeBoardMember,
  setBoardArchive,
  updateBoard,
  updateBoardMemberRole,
} from "~/server/services/board.service";
import { listBoardColumns } from "~/server/services/column.service";
import { ServiceError } from "~/server/services/errors";

import { authMiddleware } from "../middleware/auth";
import { rateLimit } from "../middleware/rate-limit";
import { jsonError, jsonList, jsonSuccess } from "../response";
import type { AuthContext } from "../types";

const boardRoutes = new Hono<AuthContext>();

const boardIdParams = z.object({
  id: z.string().min(1),
});

const boardMemberParams = z.object({
  id: z.string().min(1),
  userId: z.string().min(1),
});

const createBoardSchema = z.object({
  title: z.string().min(1).max(255),
  description: z.string().max(2000).optional(),
});

const updateBoardSchema = z.object({
  title: z.string().min(1).max(255).optional(),
  description: z.string().max(2000).nullable().optional(),
});

const archiveSchema = z.object({
  isArchived: z.boolean(),
});

const duplicateSchema = z.object({
  title: z.string().min(1).max(255).optional(),
});

const memberRoleSchema = z.enum(boardRoles);

const addMemberSchema = z.object({
  userId: z.string().min(1),
  role: memberRoleSchema,
});

const updateMemberRoleSchema = z.object({
  role: memberRoleSchema,
});

function handleServiceError(c: Context<AuthContext>, error: unknown) {
  if (error instanceof ServiceError) {
    return jsonError(c, error.message, error.status, error.code);
  }

  throw error;
}

boardRoutes.use("*", authMiddleware);
boardRoutes.use("*", rateLimit);

boardRoutes.get("/", async (c) => {
  try {
    const user = c.get("user");
    const boards = await listBoardsForUser(user.id);
    // For now, we don't have pagination implemented, so we return page 1 with all results
    return jsonList(c, boards, boards.length, 1);
  } catch (error) {
    return handleServiceError(c, error);
  }
});

boardRoutes.post("/", zValidator("json", createBoardSchema), async (c) => {
  try {
    const user = c.get("user");
    const body = c.req.valid("json");
    const result = await createBoard({
      title: body.title,
      description: body.description,
      ownerId: user.id,
    });

    return jsonSuccess(c, result, 201);
  } catch (error) {
    return handleServiceError(c, error);
  }
});

boardRoutes.get(
  "/:id",
  zValidator("param", boardIdParams),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const user = c.get("user");
      const board = await getBoardDetail(id, user.id);
      return jsonSuccess(c, board);
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.patch(
  "/:id",
  zValidator("param", boardIdParams),
  zValidator("json", updateBoardSchema),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const user = c.get("user");
      const updates = c.req.valid("json");
      const board = await updateBoard(id, user.id, updates);
      return jsonSuccess(c, { board });
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.post(
  "/:id/archive",
  zValidator("param", boardIdParams),
  zValidator("json", archiveSchema),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const { isArchived } = c.req.valid("json");
      const user = c.get("user");
      const board = await setBoardArchive(id, user.id, isArchived);
      return jsonSuccess(c, { board });
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.post(
  "/:id/duplicate",
  zValidator("param", boardIdParams),
  zValidator("json", duplicateSchema),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const user = c.get("user");
      const body = c.req.valid("json");
      const duplicated = await duplicateBoard(id, user.id, body);
      return jsonSuccess(c, duplicated, 201);
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.get(
  "/:id/columns",
  zValidator("param", boardIdParams),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const user = c.get("user");
      const columns = await listBoardColumns(id, user.id);
      return jsonSuccess(c, { columns });
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.delete(
  "/:id",
  zValidator("param", boardIdParams),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const user = c.get("user");
      await deleteBoard(id, user.id);
      return jsonSuccess(c, { deleted: true });
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.get(
  "/:id/members",
  zValidator("param", boardIdParams),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const user = c.get("user");
      const members = await listBoardMembers(id, user.id);
      return jsonSuccess(c, { members });
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.post(
  "/:id/members",
  zValidator("param", boardIdParams),
  zValidator("json", addMemberSchema),
  async (c) => {
    try {
      const { id } = c.req.valid("param");
      const { userId, role } = c.req.valid("json");
      const actor = c.get("user");
      const member = await addBoardMember(id, userId, actor.id, role);
      return jsonSuccess(c, { member }, 201);
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.patch(
  "/:id/members/:userId",
  zValidator("param", boardMemberParams),
  zValidator("json", updateMemberRoleSchema),
  async (c) => {
    try {
      const { id, userId } = c.req.valid("param");
      const { role } = c.req.valid("json");
      const actor = c.get("user");
      const member = await updateBoardMemberRole(id, userId, actor.id, role);
      return jsonSuccess(c, { member });
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

boardRoutes.delete(
  "/:id/members/:userId",
  zValidator("param", boardMemberParams),
  async (c) => {
    try {
      const { id, userId } = c.req.valid("param");
      const actor = c.get("user");
      await removeBoardMember(id, userId, actor.id);
      return jsonSuccess(c, { removed: true });
    } catch (error) {
      return handleServiceError(c, error);
    }
  },
);

export { boardRoutes };

File: api/app/routers/boards.py
from typing import Any

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import require_user
from app.responses import json_error, json_list, json_success
from app.schemas import (
    ArchiveBoardBody,
    CreateBoardBody,
    DuplicateBoardBody,
    UpdateBoardBody,
)
from app.services import board_service
from app.services.column_service import list_board_columns

router = APIRouter(prefix="/api/boards", tags=["boards"])


@router.get("")
def list_boards(
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    boards = board_service.list_boards_for_user(db, user["id"])
    return json_list(boards, len(boards), 1)


@router.post("")
def create_board(
    body: CreateBoardBody,
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    result = board_service.create_board(db, body.title, body.description, user["id"])
    return json_success(result, status=201)


@router.get("/{board_id}")
def get_board(
    board_id: str,
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    detail = board_service.get_board_detail(db, board_id, user["id"])
    return json_success(detail)


@router.patch("/{board_id}")
def update_board(
    board_id: str,
    body: UpdateBoardBody,
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    board = board_service.update_board(
        db, board_id, user["id"], body.model_dump(exclude_unset=True)
    )
    return json_success({"board": board})


@router.post("/{board_id}/archive")
def archive_board(
    board_id: str,
    body: ArchiveBoardBody,
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    board = board_service.set_board_archive(db, board_id, user["id"], body.isArchived)
    return json_success({"board": board})


@router.post("/{board_id}/duplicate")
def duplicate_board(
    board_id: str,
    body: DuplicateBoardBody,
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    detail = board_service.duplicate_board(db, board_id, user["id"], body.title)
    return json_success(detail, status=201)


@router.get("/{board_id}/columns")
def get_board_columns(
    board_id: str,
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    columns = list_board_columns(db, board_id, user["id"])
    return json_success({"columns": columns})


@router.post("/{board_id}/columns")
def create_column_disabled(
    board_id: str,
    user: dict[str, Any] = Depends(require_user),
) -> JSONResponse:
    return json_error(
        "Column creation is not available", 405, "COLUMN_CREATE_DISABLED"
    )


@router.delete("/{board_id}")
def delete_board(
    board_id: str,
    user: dict[str, Any] = Depends(require_user),
    db: Session = Depends(get_db),
) -> JSONResponse:
    board_service.delete_board(db, board_id, user["id"])
    return json_success({})
